File: java/securedoors.java
import java.util.HashSet;
import java.util.Scanner;

public class securedoors {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();
		HashSet<String> room = new HashSet<>(n);
		while (n-- > 0) {
			switch (scan.next()) {
			case "entry":
				String name1 = scan.next();
				System.out.printf("%s entered", name1);
				if (!room.add(name1))
					System.out.print(" (ANOMALY)");
				System.out.println();
				break;
			case "exit":
				String name2 = scan.next();
				System.out.printf("%s exited", name2);
				if (!room.remove(name2))
					System.out.print(" (ANOMALY)");
				System.out.println();
				break;
			}
		}
		scan.close();
	}
}

File: ts/src/securedoors.ts
import readline = require('readline')

enum DoorState {
  Entry = 'entry',
  Exit = 'exit',
}

const room: Set<string> = new Set()

let lineCount = 0

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount > 0) {
      const [state, name] = line.split(' ')
      if (state === DoorState.Entry) {
        if (room.has(name ?? '')) {
          console.log(`${name} entered (ANOMALY)`)
        } else {
          room.add(name ?? '')
          console.log(`${name} entered`)
        }
      } else if (state === DoorState.Exit) {
        if (room.has(name ?? '')) {
          room.delete(name ?? '')
          console.log(`${name} exited`)
        } else {
          console.log(`${name} exited (ANOMALY)`)
        }
      }
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
